proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,1047 @@
+
+This is not done automatically in order to avoid confusion around
+existential varibles.
+
+Explicit specificity
+~~~~~~~~~~~~~~~~~~~~
+
+When supplying type arguments to matchability-polymorphic functions such as ::
+
+   qux :: forall m (f :: Type -> @m Type) a. f a -> f a
+
+the user needs to provide either a concrete matchability or a wildcard before
+supplying the instantiation for ``f``, as in ``qux @_ @Id``. This is tiresome,
+because ``m`` can *always* be inferred from the kind of ``f``, so it would be
+preferable to write ``qux @Id`` instead.
+
+The `explicit specificity <https://github.com/ghc-proposals/ghc-proposals/pull/99>`_
+feature greatly improves the usability of unsaturated type families, as now the signature
+can be written as ::
+
+   qux :: forall {m} (f :: Type -> @m Type) a. f a -> f a
+
+Linear Haskell
+~~~~~~~~~~~~~~
+
+Under ``LinearTypes``, the arrow type is decorated with a different kind of
+information: multiplicity. Other than syntactic considerations and somewhat
+overlapping implementations, there is no interaction between matchability and
+multiplicity.
+
+Comparison with levity polymorphism
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here we draw a comparsion between matchability polymorphism and `levity polymorphism <https://www.microsoft.com/en-us/research/wp-content/uploads/2016/11/levity-pldi17.pdf>`_,
+from the perspective of type inference. There is no notable interaction between these
+two features, but there are noteworthy differences between the way matchability variables
+are inferred compared to runtime representation variables.
+
+In *types*, runtime representation variables are all defaulted to ``LiftedRep``, and
+matchability variables are all defaulted depending on where the variables appear
+(see the *Term-level functions* and *Kind-arrows in type signatures* sections
+above).
+
+In *kinds*, runtime representation variables are all defaulted to ``LiftedRep``,
+but matchability variables are only defaulted when a signature is given, and
+generalised otherwise.
+
+As a simple example, consider the following two type families ::
+
+  type Foo :: forall {r :: RuntimeRep} {m :: Matchability}. TYPE r -> @m TYPE r
+  type Foo = ...
+
+  -- inferred: Bar :: forall {m :: Matchability}. Type -> @m Type
+  type Bar = Foo
+
+``Foo`` is both levity-polymorphic and matchability-polymorphic. However, in
+``Bar``'s kind, the ``RuntimeRep`` variable is defaulted, but the
+``Matchability`` variable is generalised.
+
+The rationale behind defaulting runtime rep variables in types is that inferring
+polymorphism would trip up code generation. The rationale behind defaulting
+matchabilities in types is that inferring polymorphism would lead to ambiguous
+types. In kinds, however, we take a more nuanced approach, because
+generalisation there is desirable.
+
+See the *Higher-order arguments* section above and the *Alternatives* section below for
+more details behind this approach.
```

Comment:
RST lets you add hyperlinks to other sections, please do so: it's very helpful.

Reply:
I've tried that before, but github's viewer didn't generate the crossref targets unfortunately. Would it be worth it still to add hyperlinks?